### PR TITLE
CDAP-12328 Cache visibility response in Remote Authorizer

### PR DESCRIPTION
**Pull request by poornachandra**
JIRA - https://issues.cask.co/browse/CDAP-12328
Build - https://builds.cask.co/browse/CDAP-RUT1241-3

### CODE REVIEW

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java [hunk 1]
```diff
@@ -131,44 +139,68 @@ public void testPrivilegesManager() throws Exception {
...
     privilegesManager.grant(NS, ALICE, EnumSet.allOf(Action.class));
     privilegesManager.grant(APP, ALICE, Collections.singleton(Action.ADMIN));
     privilegesManager.grant(PROGRAM, ALICE, Collections.singleton(Action.EXECUTE));
     authorizationEnforcer.enforce(NS, ALICE, EnumSet.allOf(Action.class));
     authorizationEnforcer.enforce(APP, ALICE, Action.ADMIN);
     authorizationEnforcer.enforce(PROGRAM, ALICE, Action.EXECUTE);
-    Map<AuthorizationPrivilege, Boolean> authorizationPrivilegeBooleanMap =
-      ((RemoteAuthorizationEnforcer) authorizationEnforcer).cacheAsMap();
-    // 4 For EnumSet.allOf(Action.class) and one each for the next 2
-    Assert.assertEquals(currentCount + 6, authorizationPrivilegeBooleanMap.size());
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:144]: Shall we add these to the test with cache enabled?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      // Expected
+    }
+  }
+
+  @Override
+  public void testVisibility() throws Exception {
+    super.testVisibility();
+
+    // The super class revokes all privileges after test is done. Since cache is disabled, nothing should be visible.
+
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java]: extra new line

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    Assert.assertEquals(ImmutableSet.of(),
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java]: We are do `Assert.assertTrue(authorizationEnforcer.isVisible(ImmutableSet.of(NS, APP, PROGRAM), ALICE).isEmpty())`

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    authorizationEnforcer.enforce(APP, ALICE, Action.ADMIN);
+    authorizationEnforcer.enforce(PROGRAM, ALICE, Action.EXECUTE);
+  }
+
+  @Override
+  public void testVisibility() throws Exception {
+    super.testVisibility();
+
+    // The super class revokes all privileges after test is done. Since cache is enabled, visibility should still work.
+
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesCacheTest.java]: extra new line

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java [hunk 5]
```diff
@@ -131,44 +139,68 @@ public void testPrivilegesManager() throws Exception {
...
+    // Grant privileges on entities ending with 2 to BOB
+    privilegesManager.grant(program2, BOB, Collections.singleton(Action.ADMIN));
+    privilegesManager.grant(ds2, BOB, EnumSet.of(Action.READ, Action.WRITE));
+    privilegesManager.grant(stream2, BOB, EnumSet.allOf(Action.class));
+
+    Set<? extends EntityId> allEntities = ImmutableSet.of(NS,
+                                                          APP, PROGRAM, ds, stream,
+                                                          app1, program1, ds1, stream1,
+                                                          app2, program2, ds2, stream2
+    );
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java]: We can put the parentheses to the previous line.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java [hunk 6]
```diff
@@ -131,44 +139,68 @@ public void testPrivilegesManager() throws Exception {
...
+
+    Assert.assertEquals(ImmutableSet.of(NS, APP, PROGRAM, ds, stream),
+                        authorizationEnforcer.isVisible(allEntities, ALICE));
+
+    Assert.assertEquals(ImmutableSet.of(NS, app2, program2, ds2, stream2),
+                        authorizationEnforcer.isVisible(allEntities, BOB));
+
+    Assert.assertEquals(ImmutableSet.of(),
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:191]: We can do `Assert.assertTrue(authorizationEnforcer.isVisible(allEntities, CAROL).isEmpty())`

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java [hunk 7]
```diff
@@ -131,44 +139,68 @@ public void testPrivilegesManager() throws Exception {
...
+                        authorizationEnforcer.isVisible(allEntities, CAROL));
+
+    Assert.assertEquals(ImmutableSet.of(),
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:194]: We can do `Assert.assertTrue(authorizationEnforcer.isVisible(ImmutableSet.<EntityId>of(), ALICE).isEmpty())` here

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Test {@link RemoteAuthorizationEnforcer} with cache disabled.
+ */
+public class RemotePrivilegesNoCacheTest extends RemotePrivilegesTestBase {
+
+  @BeforeClass
+  public static void beforeClass() throws IOException, InterruptedException {
+    cConf.setInt(Constants.Security.Authorization.CACHE_MAX_ENTRIES, 0);
+    RemotePrivilegesTestBase.setup();
+
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java]: extra new line

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java [hunk 9]
```diff
@@ -62,12 +70,30 @@
   private static final Gson GSON = new GsonBuilder()
     .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
     .create();
-  private static final Type SET_ENTITY_TYPE = new TypeToken<Set<EntityId>>() { }.getType();
+  private static final Type SET_ENTITY_TYPE = new TypeToken<Set<EntityId>>() {
+  }.getType();
```
Comment (yaojiefeng) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java]: We can put this in the previous line

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java [hunk 10]
```diff
@@ -119,15 +183,82 @@ private boolean doEnforce(AuthorizationPrivilege authorizationPrivilege) throws
...
 
-  @VisibleForTesting
-  public Map<AuthorizationPrivilege, Boolean> cacheAsMap() {
-    return Collections.unmodifiableMap(authPolicyCache.asMap());
+  private Map<VisibilityKey, Boolean> loadVisibility(Iterable<? extends VisibilityKey> keys) throws IOException {
+    if (!keys.iterator().hasNext()) {
+      return Collections.emptyMap();
+    }
+
+    Principal principal = keys.iterator().next().getPrincipal();
```
Comment (yaojiefeng) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java]: Shall we put a comment saying by the current implementation, the principal will not be a set?

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java [hunk 11]
```diff
@@ -119,15 +183,82 @@ private boolean doEnforce(AuthorizationPrivilege authorizationPrivilege) throws
...
+    }
+    return keyMap;
+  }
+
+  private Set<? extends EntityId> toEntityIds(Iterable<? extends VisibilityKey> keys) {
+    return ImmutableSet.copyOf(Iterables.transform(keys, VISIBILITY_KEY_ENTITY_ID_FUNCTION));
+  }
+
+  private Iterable<VisibilityKey> toVisibilityKeys(final Principal principal, Set<? extends EntityId> entityIds) {
+    return Iterables.transform(entityIds, new Function<EntityId, VisibilityKey>() {
```
Comment (yaojiefeng) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:210]: Shall we also make this Function a global variable?

**Review by yaojiefeng — COMMENTED** [review]
Just some minor comments

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java (hunk 1 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:144]: We are not testing `LoadingCache`, so no point in doing those asserts

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java (hunk 6 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:191]: I think having `(expected, actual)` helps in readability

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java (hunk 7 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesTestBase.java:194]: Same here about redability

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java (hunk 3 above)
Comment (poornachandra) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/remote/RemotePrivilegesNoCacheTest.java]: Same here about redability

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java (hunk 11 above)
Comment (poornachandra) [reply cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java:210]: It needs access to a local variable, hence left it here.

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java (hunk 10 above)
Comment (poornachandra) [reply cdap-security/src/main/java/co/cask/cdap/security/authorization/RemoteAuthorizationEnforcer.java]: Good point

**Review by yaojiefeng — APPROVED** [review]
LGTM

**Comment by poornachandra** [comment]
Build is green, will squash and merge.

**Pull request merged**